### ISSUE TITLE
Make blocking operations asynchronous

### ISSUE THREAD

**Issue by jantic** [lanpa/tensorboardX#372]
Hello-  I've recently integrated tensorboardX with the core fastai library.  One thing that I needed to work around in order to get acceptable performance was to put blocking operations behind a separate thread so that the rest of training would not be impeded.  The two big operations causing the need for this were histograms (acknowledged already I can see) and writing images.  

Writing images appears to block on this line in def make_image, summary.py:

`image.save(output, format='PNG') (in def make_image in summary.py).`

I determined this using cProfile and snakeviz.

Histograms appear to be bottlenecked at numpy construction.

Would it be possible for TensorboardX to do the work of queuing up these sorts of bottle-necking requests and hiding the processing via another thread or process? It seems like this would be a broadly useful thing to do, and it would definitely make an impact on the user experience.

**Comment by lanpa** [comment]
Yes, they are CPU intensive task. I have considered this before and found that making `make_image` async might cause the `global iteration` in tensorboard interleaved. (while an big image of iter=5 is under processing, an `add_scalar` with iter=6 is called) This makes tensorboard server unhappy and drop one of them. I am not sure if this behaviour is still there. Will check this weekend. If the global iteration in the event file must be monotonically increased, then the bound might still be on the IO side. Another solution I can think of is making a BIG buffer to queue the events and make sure that `global iteration` meets tensorboard requirement. (ex. queue all `iter>5` before all `iter=5` is written to disk)

**Comment by jantic** [comment]
Thanks!

**Comment by egm108** [comment]
@lanpa, could you please clarify about your project tensorboardX?
Do I understand correctly the tensorboardX history?
1) you had developed tensorboardX, then it has been added to pytorch. So there is 2 active branches of tensorboardX: here and in torch.utils.
2)you had requested that tensorboard should be separated from tensorflow. And they are separate now.
So what I didn't understand:
1)are you going support both versions (here and pytorch)?
2) did you use initially tensorflow's code? I know there is, say EventFileWriter in tensorboard from you, right? And also it asynchronous, right? So can we expect heavy operations in separate thread?
So what is tensorboard (separated from tensorflow) now? is this code by you or by tensorflow?

Thank you!
